operator/src/main/java/org/bf2/operator/operands/KafkaCluster.java

```diff
@@ -479,7 +479,7 @@ private KafkaClusterTemplate buildKafkaTemplate(ManagedKafka managedKafka) {
         KafkaClusterTemplateBuilder templateBuilder = new KafkaClusterTemplateBuilder()
                 .withPod(podTemplateBuilder.build());
 
-        if (drainCleanerManager.isDrainCleanerWebhookFound()) {
+        if (replicas > 1 && drainCleanerManager.isDrainCleanerWebhookFound()) {
             templateBuilder.withPodDisruptionBudget(
                 new PodDisruptionBudgetTemplateBuilder()
                     .withMaxUnavailable(0)
```

Comment:
More of a question, 
1) why is this enabled only when drain cleaner webhook is found?
2) with `maxUnavilable` of `0` it says all the broker nodes must be available. I wonder how the OCP upgrades are taking place, I thought should this be the same as insync-replica count?

Reply:
Testing my understanding of this :) ... having it set to `0` means that we rely on the presence of the drain cleaner to label the pods for Strimzi to restart the pods.

Reply:
@MikeEdgar yes that is correct. 

@rareddy see https://github.com/strimzi/drain-cleaner#see-it-in-action

Reply:
Thanks for explanation

Reply:
We set `PDB=0` for Kafka and ZK pods in order to block K8s draining and delegate this task to the combined work of the Drain Cleaner and the Cluster Operator, **regardless of the number of replicas**. This is to avoid service disruption, as the CO knows when a replica is fully synced and we can drain the next one. 

@shawkins this change should be also applied to `buildZookeeperTemplate`, but I think it is not needed. There are corner cases in which it is never safe to roll a broker. A single broker cluster is one, a topic with `RF < min.insync.replicas` is another. The CO is smart enough to recognize these cases and proceed with the broker roll. This is also confirmed by a quick test that I did using the latest version of the DC and the CO.

Reply:
@fvaleri When you say that it's not needed, do you mean that the auto-scaler will still be able to reschedule developer instances away from nodes that are under their utilization threshold - https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#does-ca-work-with-poddisruptionbudget-in-scale-down

Because these are developer instances cost concerns trump safety.

Reply:
Not exactly, I mean that both the drain command and the cluster-autoscaler will start retrying periodically because of the PDB settings. In the meantime, DC+CO will do their work moving out any ZK and Kafka pod from the terminating node. This means the node drain will eventually succeed. For the drain command the default timeout is infinite, while it looks like the CA waits at most 2 minutes, which is probably not be enough and should be raised based on testing.

Reply:
My reading of the autoscaler source is that prior to attempting the draining, it will do a quick check of if the pdbs are satiable:

https://github.com/kubernetes/autoscaler/blob/5745044ddf4378c1aae16c47fb173b1abf709c8e/cluster-autoscaler/simulator/drain.go#L102

So I think just having the pdb on the developer instances is enough to block the auto-scaling attempt at draining.

> while it looks like the CA waits at most 2 minutes, which is probably not be enough and should be raised based on testing.

Unfortunately we have almost no settings that can be tuned via the openshift autoscaling - for example see https://issues.redhat.com/browse/HIVE-1976 - if a timeout change is needed, it will require a change pretty far upstream.

Reply:
I see, but this is not how the node drain should work. I can understand that the autoscaler can't wait indefinitely, but it should at least be configurable.

The problem with this change is that the Drain Cleaner would still be notified and the cluster will still be rolled. This may not be a problem for a developer instance, but it's not desired for bigger instances.

Reply:
> I see, but this is not how the node drain should work.  I can understand that the autoscaler can't wait indefinitely, but it should at least be configurable.

There would need to be two changes here correct - one change to allow draining to even be attempted when disruptionsAllowed=0, and another to allow for the draining timeout to be configurable.

> The problem with this change is that the Drain Cleaner would still be notified and the cluster will still be rolled.

There are two cases here - staging clusters will have both developer and standard instances.  While production clusters should have only a single type.  Would you advocate in the longer term for production developer clusters to have the drain cleaner webhook not installed?

> but I guess it's not desired for bigger instances.

This change will only affect developer instances.  The standard plan will always have more than a single replica.

Reply:
> There are two cases here - staging clusters will have both developer and standard instances. While production clusters should have only a single type. 

The OLM bundle should be smart enough to know if the drain cleaner have to be deployed or not based on the environment type. I don't think it has this capability, so we would need a new lightweight dev-prod bundle with no multi version and drain cleaner deployments.

> Would you advocate in the longer term for production developer clusters to have the drain cleaner webhook not installed?

Yes. The DC+CO draining will still work, but with single-broker instances there would be no benefit comparing to the standard K8s draining process.